### PR TITLE
[Bug fix] Handle -0.0 predicate correctly in ttnn.where float32 path

### LINKED ISSUE #56522: `ttnn.where` (float32): a `-0.0` predicate selects the *true* branch — the SFPU kernel tests the raw bit pattern against zero, so `-0.0 != 0`, where torch and ttnn's own golden (`torch.ne(predicate, 0)`) select the false branch
Verified against tt-metal `main` @ `70f840f46adb33ec21ec8893f48ee06692648f75`, Blackhole ttsim v1.10.7, SFPI 7.76.0.
Evidence, probe and recorded rows: https://gist.github.com/beyondpremiere/172dc3264cb99ee638c3255e74289aa3

---

**Component / Area:** SFPU numerics — `tt_metal/tt-llk/tt_llk_{blackhole,wormhole_b0}/common/inc/sfpu/ckernel_sfpu_where.h`
**Issue Type:** Correctness

## Summary

`_calculate_where_` loads the predicate lane as a raw 32-bit integer (`mod0 = InstrModLoadStore::INT32`
for Float32, `ckernel_sfpu_where.h:29`) and decides the branch with

```
TTI_SFPSETCC(0, p_sfpu::LREG0, 0, sfpi::SFPSETCC_MOD1_LREG_EQ0);   // ckernel_sfpu_where.h:37 (and the
                                                                   //  SFPLOADMACRO variants at :54, :80)
```

That is an integer `== 0` on the bit pattern. `-0.0` is `0x80000000`, so it is classified as non-zero
and the kernel returns `true_value`. IEEE 754 says `-0.0 == 0`; `torch.where(cond.bool(), ...)` treats
`-0.0` as `False`; and ttnn's own reference does too — `ttnn/ttnn/operations/ternary.py:73,89`
normalises the predicate with `torch.ne(predicate, 0)` before calling `torch.where`. The op therefore
disagrees with its own golden on this input.

`ttnn.where` passes a float predicate straight to the kernel (`ternary.cpp:139-144` only typecasts the
dtype when it differs from the value tensors; there is no `!= 0` normalisation), so this reaches users
directly. Related to #43831 (same sign-bit-vs-zero pattern in the zero-comparison ops, fixed there) and
#56520 (`logical_not`, same residual).

## Observed

`sfpu_ternary_test.cpp`, `SfpuWhere`, Float32 in/out, `dest_acc=Yes`; predicate tile of 19 crafted
values, `true_value = 1000 + i`, `false_value = -(1000 + i)`, compared against
`torch.where(pred.bool(), a, b)`:

| predicate | bits | kernel picks | torch picks |
|---|---|---|---|
| `+0.0` | `0x00000000` | false | false |
| **`-0.0`** | `0x80000000` | **true** | false |
| `±1`, `±0.5`, `2`, `±1e-20`, `FLT_MAX` | — | true | true |
| `±inf` | — | true | true |
| NaN (`0x7fc00000`, `0xffc00000`, `0x7fc00001`) | — | true | true |
| `±min normal`, `±smallest subnormal` | — | true | true |

Only `-0.0` disagrees. `-0.0` is an ordinary product of float arithmetic (`0.0 * -x`, `-x` of a zero, a
negative product underflowing), so a predicate built from such a tensor silently flips those lanes.

## Proposed fix

For float layouts, drop the sign bit before the zero test. One instruction, no constant register:
shifting the pattern left by one turns `0x80000000` into `0` and leaves every non-zero magnitude
non-zero.

```cpp
// DISABLE_SFPLOADMACRO path, Float32 layout
TT_SFPLOAD(p_sfpu::LREG0, mod0, ADDR_MOD_7, offset0);
TT_SFPLOAD(p_sfpu::LREG1, mod0, ADDR_MOD_7, offset1);
TTI_SFPSHFT(1, p_sfpu::LREG0, p_sfpu::LREG0, 1);   // LREG0 <<= 1: -0.0 -> 0, everything else stays non-zero
TTI_SFPSETCC(0, p_sfpu::LREG0, 0, sfpi::SFPSETCC_MOD1_LREG_EQ0);
```

(replay length 6 → 7). **Validated on the simulator:** with that change, all 19 predicates above match
torch, `-0.0` included, and nothing else moves (`patch_where_signed_zero.py`). The `SFPLOADMACRO`
variants need the same shift inserted between the predicate load and the `SFPSETCC`. For the
`Float16_b` layout (`LO16`) the sign bit is bit 15, so the equivalent is a mask of `0x7FFF` rather
than a shift; I measured the Float32 path only. The Wormhole copy differs from Blackhole only in
`ADDR_MOD` constants and needs the same edit.

## Reproduction

`probe_where.py` (drop into `tt_metal/tt-llk/tests/python_tests/`, run with `--run-simulator --forked`,
`CHIP_ARCH=blackhole`). `baseline_where_specials.csv` and `fixed_where_specials.csv` are the recorded
outputs (predicate bits, kernel output, torch).

## Scope

Simulator-only (Blackhole ttsim). The mechanism is a bit-pattern compare in the kernel source, not a
simulator behaviour, but the result should be confirmed on hardware.

### PR DESCRIPTION
**Pull request by shitobu**
### PR Category
Bug fix

### Summary
Fixes #56522

`_calculate_where_` in `ckernel_sfpu_where.h` tests the predicate lane against zero using `SFPSETCC_MOD1_LREG_EQ0` (integer == 0 test). For Float32, IEEE 754 `-0.0` is `0x80000000`, so it is classified as non-zero and selects the `true_val` branch instead of `false_val`. In contrast, IEEE 754, `torch.where(cond.bool(), ...)`, and `ttnn`'s golden reference (`torch.ne(predicate, 0)`) treat `-0.0` as `false`.

#### Changes:
1. **Blackhole & Wormhole B0 (`ckernel_sfpu_where.h`)**:
   - For `data_format == DataFormat::Float32`, shift the predicate register left by 1 (`TTI_SFPSHFT(1, p_sfpu::LREG0, p_sfpu::LREG0, 1)`) before `TTI_SFPSETCC`. Shifting left drops bit 31 (the sign bit) so `0x80000000` becomes `0`, while all non-zero float values retain non-zero bits.
   - For non-Float32 data formats (`Int32`, `UInt32`), the existing fast macro path remains untouched.
2. **Quasar (`ckernel_sfpu_where.h`)**:
   - Compares on `sfpi::abs(cond) != 0.0f` instead of raw `cond != 0` so that `-0.0f` matches zero.
3. **Tests (`test_where.py`)**:
   - Added `test_ttnn_where_signed_zero_predicate` covering both `+0.0` and `-0.0` predicates against torch reference.

### Notes for reviewers
- The shift left by 1 instruction is 1 cycle and requires no additional constant registers.
- All non-Float32 paths (Int32, UInt32, Float16_b) and non-where ops remain completely unaffected.

### CODE REVIEW

**Comment by ncvetkovicTT** [comment]
@mateusznowakTT FYI.

File: tests/ttnn/unit_tests/operations/eltwise/test_where.py [hunk 1]
```diff
@@ -87,6 +87,25 @@ def test_ttnn_where(c_shape, t_shape, f_shape, scalar, variant, condition, devic
     assert torch_equal_nan(result, golden)
 
 
+def test_ttnn_where_signed_zero_predicate(device):
+    # Tests that both +0.0 and -0.0 predicates correctly evaluate to false branch (closes #56522)
+    torch.manual_seed(0)
+    shape = (1, 1, 32, 32)
+    c_data = torch.tensor([0.0, -0.0, 1.0, -1.0], dtype=torch.float32).repeat(32 * 32 // 4).reshape(shape)
```
Comment (mateusznowakTT) [inline tests/ttnn/unit_tests/operations/eltwise/test_where.py]: Please extend the test also to bf16 and make sure it is working properly there.

File: tt_metal/tt-llk/tt_llk_wormhole_b0/common/inc/sfpu/ckernel_sfpu_where.h [hunk 2]
```diff
@@ -42,10 +48,30 @@ inline void _calculate_where_(
...
     }
 #else
-    if (dst_index_out == dst_index_in0)
+    if constexpr (data_format == DataFormat::Float32)
+    {
+        // For Float32, shift left by 1 to discard the sign bit before testing for zero (IEEE 754: -0.0 == 0).
+        int offset3 = (dst_index_out * 32) << 1;
+
+        lltt::record(0, 7);
+        TT_SFPLOAD(p_sfpu::LREG0, mod0, ADDR_MOD_3, offset0);
```
Comment (mateusznowakTT) [inline tt_metal/tt-llk/tt_llk_wormhole_b0/common/inc/sfpu/ckernel_sfpu_where.h]: This change costs +25% on perf.
Please try to reuse SFPLOADMACROs to achieve the result (I believe this should be doable with 1 more step vs. the current pipeline).

**Review by mateusznowakTT — CHANGES_REQUESTED** [review]

File: tests/ttnn/unit_tests/operations/eltwise/test_where.py (hunk 1 above)
Comment (shitobu) [reply tests/ttnn/unit_tests/operations/eltwise/test_where.py]: Done! `test_ttnn_where_signed_zero_predicate` is now parametrized over both `torch.float32`/`ttnn.float32` and `torch.bfloat16`/`ttnn.bfloat16`, verifying that signed zero predicates evaluate to the false branch on both formats.

File: tt_metal/tt-llk/tt_llk_wormhole_b0/common/inc/sfpu/ckernel_sfpu_where.h (hunk 2 above)
Comment (shitobu) [reply tt_metal/tt-llk/tt_llk_wormhole_b0/common/inc/sfpu/ckernel_sfpu_where.h]: Updated in commit `accad9d`! Restored and reused the `SFPLOADMACRO` pipeline with exactly +1 step as suggested (4 cycles for in-place, 5 cycles for out-of-place):

1. **Pipeline configuration in `_init_where_`**:
   - Macro 0 & Macro 1 `simple_bits`: configured with `(1 << 3) | 4` (delay 1 for `SFPSETCC`, firing 2 instructions later).
   - Macro 0 `store_bits`: configured with `(3 << 3) | 3` (delay 3, firing store 4 instructions later at the start of the next iteration for in-place).
2. **Cycle timing in `_calculate_where_`**:
   - **Cycle 0**: `TT_SFPLOADMACRO(macro0/1, ...)` loads condition into `L0` (schedules `SFPSETCC` with delay 1).
   - **Cycle 1**: `TTI_SFPSHFT(shift, L0, L0, 1)` clears sign bit (`shift = 1` for `Float32`, `shift = 17` for `Float16_b`, `TTI_SFPNOP` for integer).
   - **Cycle 2**: `TT_SFPLOADMACRO(macro2, ...)` loads `true_val` into `L0`, while Simple Unit executes queued `SFPSETCC LaneEnabled=(L0 EQ 0)` on the shifted condition.
   - **Cycle 3**: `TT_SFPLOAD` loads `false_val` into `L0` predicated on `LaneEnabled`, while Simple Unit executes queued `SFPENCC (LaneEnabled=true)`.
   - **Cycle 4**: Out-of-place store (for in-place, store triggers at Cycle 0 of subsequent iteration).

This eliminates the unpipelined replay loop and brings perf back to the optimal 4/5 cycle pipeline across both Wormhole B0 and Blackhole.

**Comment by mateusznowakTT** [comment]
Ok, recent change +5.8% on perf (+0.03 cycles/datum, fp32 only), functionally correct.

**Review by mateusznowakTT — APPROVED** [review]

**Comment by ldjurovicTT** [comment]
Please just make it visible in tt-llk/tests/python_tests in ttnn where's test that this special case is resolved

**Comment by shitobu** [comment]
Thanks @ldjurovicTT! Updated in `e95a53b2`:
- Added `"all_negative_zeros"` (full multi-tile condition tensor filled with `-0.0`) and `"signed_zeros"` (alternating `[0.0, -0.0, 1.0, -1.0]` condition pattern) directly to `test_ttnn_where` in `tt_metal/tt-llk/tests/python_tests/test_sfpu_ternary.py` as well as the Quasar suite (`test_sfpu_where_quasar.py`).
- Verified these test cases explicitly validate against `WhereGolden` that `-0.0` evaluates to the false branch as expected on SFPU.
